src/smallmap_gui.cpp

```diff
@@ -1225,40 +1265,46 @@ class SmallMapWindow : public Window {
 		StringID legend_tooltip;
 		StringID enable_all_tooltip;
 		StringID disable_all_tooltip;
+		bool disable_industry_names;
 		int plane;
 		switch (this->map_type) {
 			case SMT_INDUSTRY:
 				legend_tooltip = STR_SMALLMAP_TOOLTIP_INDUSTRY_SELECTION;
 				enable_all_tooltip = STR_SMALLMAP_TOOLTIP_ENABLE_ALL_INDUSTRIES;
 				disable_all_tooltip = STR_SMALLMAP_TOOLTIP_DISABLE_ALL_INDUSTRIES;
+				disable_industry_names = false;
 				plane = 0;
 				break;
 
 			case SMT_OWNER:
 				legend_tooltip = STR_SMALLMAP_TOOLTIP_COMPANY_SELECTION;
 				enable_all_tooltip = STR_SMALLMAP_TOOLTIP_ENABLE_ALL_COMPANIES;
 				disable_all_tooltip = STR_SMALLMAP_TOOLTIP_DISABLE_ALL_COMPANIES;
+				disable_industry_names = true;
 				plane = 0;
 				break;
 
 			case SMT_LINKSTATS:
 				legend_tooltip = STR_SMALLMAP_TOOLTIP_CARGO_SELECTION;
 				enable_all_tooltip = STR_SMALLMAP_TOOLTIP_ENABLE_ALL_CARGOS;
 				disable_all_tooltip = STR_SMALLMAP_TOOLTIP_DISABLE_ALL_CARGOS;
+				disable_industry_names = true;
 				plane = 0;
 				break;
 
 			default:
 				legend_tooltip = STR_NULL;
 				enable_all_tooltip = STR_NULL;
 				disable_all_tooltip = STR_NULL;
+				disable_industry_names = true;
 				plane = 1;
 				break;
 		}
 
 		this->GetWidget<NWidgetCore>(WID_SM_LEGEND)->SetDataTip(STR_NULL, legend_tooltip);
 		this->GetWidget<NWidgetCore>(WID_SM_ENABLE_ALL)->SetDataTip(STR_SMALLMAP_ENABLE_ALL, enable_all_tooltip);
 		this->GetWidget<NWidgetCore>(WID_SM_DISABLE_ALL)->SetDataTip(STR_SMALLMAP_DISABLE_ALL, disable_all_tooltip);
+		this->GetWidget<NWidgetCore>(WID_SM_SHOW_IND_NAMES)->SetDisabled(disable_industry_names);
```

Comment:
I believe the button should only be shown with `SMT_INDUSTRY` map type (not just disabled).

From the existing code, it looks like a `NWidgetStacked` container is the typical way to show/hide something.

I tried wrapping the toggle with a `NWID_SELECTION`, but it looks like `NWID_SELECTION` is not supposed to be stacked (crashed):

```
NWidget(NWID_SELECTION, INVALID_COLOUR, WID_SM_SHOW_IND_NAMES_CONTAINER),
  NWidget(WWT_TEXTBTN, COLOUR_BROWN, WID_SM_SHOW_IND_NAMES), SetDataTip(STR_SMALLMAP_SHOW_INDUSTRY_NAMES, STR_SMALLMAP_TOOLTIP_SHOW_INDUSTRY_NAMES),
EndContainer(),
```

I tried using `NWID_HORIZONTAL` too but other usages don't use a `WID_` with it.

What would be the best way to show/hide the toggle button?

Reply:
I agree that the button should be hidden in other map types. You are on the right track despite the crash: see `WID_BV_BUILD_SEL` in `build_vehicle_gui.cpp`. 🙂